nexus/db-queries/src/db/queries/region_allocation.rs

```diff
@@ -77,611 +64,344 @@ pub fn from_diesel(e: DieselError) -> external::Error {
+      (dataset.kind = 'crucible')
+    )
+    ORDER BY dataset.pool_id, md5((CAST(dataset.id as BYTEA) || ").param().sql("))
+  ),")
+    .bind::<sql_types::Bytea, _>(seed.clone())
+    // We order by md5 to shuffle the ordering of the datasets.
+    // md5 has a uniform output distribution so it does the job.
+    .sql("
+  shuffled_candidate_datasets AS (
+    SELECT
+      candidate_datasets.id,
+      candidate_datasets.pool_id
+    FROM candidate_datasets
+    ORDER BY md5((CAST(candidate_datasets.id as BYTEA) || ").param().sql(")) LIMIT ").param().sql("
+  ),")
+    .bind::<sql_types::Bytea, _>(seed)
+    .bind::<sql_types::BigInt, _>(redundancy)
+    // Create the regions-to-be-inserted for the volume.
+    .sql("
+  candidate_regions AS (
+    SELECT
+      gen_random_uuid() AS id,
+      now() AS time_created,
+      now() AS time_modified,
+      shuffled_candidate_datasets.id AS dataset_id,
+      ").param().sql(" AS volume_id,
+      ").param().sql(" AS block_size,
+      ").param().sql(" AS blocks_per_extent,
+      ").param().sql(" AS extent_count
+    FROM shuffled_candidate_datasets
+  ),")
+    .bind::<sql_types::Uuid, _>(volume_id)
+    .bind::<sql_types::BigInt, _>(block_size as i64)
+    .bind::<sql_types::BigInt, _>(blocks_per_extent as i64)
+    .bind::<sql_types::BigInt, _>(extent_count as i64)
+    // A subquery which summarizes the changes we intend to make, showing:
+    //
+    // 1. Which datasets will have size adjustments
+    // 2. Which pools those datasets belong to
+    // 3. The delta in size-used
+    .sql("
+  proposed_dataset_changes AS (
+    SELECT
+      candidate_regions.dataset_id AS id,
+      dataset.pool_id AS pool_id,
+      ((candidate_regions.block_size * candidate_regions.blocks_per_extent) * candidate_regions.extent_count) AS size_used_delta
+    FROM (candidate_regions INNER JOIN dataset ON (dataset.id = candidate_regions.dataset_id))
+  ),")
+    // Confirms whether or not the insertion and updates should
+    // occur.
+    //
+    // This subquery additionally exits the CTE early with an error if either:
+    // 1. Not enough datasets exist to provision regions with our required
+    //    redundancy, or
+    // 2. Not enough space exists on zpools to perform the provisioning.
+    //
+    // We want to ensure that we do not allocate on two datasets in the same
+    // zpool, for two reasons
+    // - Data redundancy: If a drive fails it should only take one of the 3
+    //   regions with it
+    // - Risk of overallocation: We only check that each zpool as enough
+    //   room for one region, so we should not allocate more than one region
+    //   to it.
+    //
+    // Selecting two datasets on the same zpool will not initially be
+    // possible, as at the time of writing each zpool only has one dataset.
+    // Additionally, we intend to modify the allocation strategy to select
+    // from 3 distinct sleds, removing the possibility entirely. But, if we
```

Comment:
This modification has been done already right?

Reply:
I'll update the wording -- it has been done, but as a configuration option, so it's not *always* turned on